components/sup/src/manager/user_config_watcher.rs

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::thread::Builder as ThreadBuilder;
+
+use super::file_watcher::{Callbacks, default_file_watcher};
+
+use hcore::fs::USER_CONFIG_FILE;
+use hcore::service::ServiceGroup;
+use manager::service::Service;
+
+static LOGKEY: &'static str = "UCM";
+
+// This trait exists to ease the testing of functions that receive a Service.  Creating Services
+// requires a lot of ceremony, so we work around this with this trait.
+pub trait Serviceable {
+    fn name(&self) -> &String;
+    fn path(&self) -> &PathBuf;
+    fn service_group(&self) -> &ServiceGroup;
+}
+
+impl Serviceable for Service {
+    fn name(&self) -> &String {
+        &self.pkg.name
+    }
+
+    fn path(&self) -> &PathBuf {
+        &self.pkg.svc_path
+    }
+
+    fn service_group(&self) -> &ServiceGroup {
+        &self.service_group
+    }
+}
+
+
+pub struct UserConfigWatcher {
+    // TODO use a HashSet? we don't really need the boolean:
```

Comment:
This comment does not seem to be true - `reset_events_for` sets the boolean to false.

Reply:
Yes, this is a note to myself saying that we could change the implementation. True/false values can be represented by presence/absence in a set.

Reply:
Tried using a `HashSet` and turns out it's more complicated than expected (if we put a `HashSet` in an `Arc`, that `HashSet` is immutable, unless I use mutexes), so not worth it.

Reply:
Yeah, you would need to do something similar to what some code in butterfly does - `Arc<RwLock<HashSet<Whatever>>>`. Anyway, I think that current solution is fine as well.